### PR TITLE
role-registry + granulare EFNM call whitelist

### PR DESCRIPTION
**Pull request by solipsis**
This PR adds a RoleRegistry contract that can serve as the central point of permissions for future updates. It is bundled with an update to switch the EtherfiNodesManager blacklist to a more secure and granular whitelist implementation. I bundled these changes to ensure the role registry was capable of fulfilling its first intended purpose

This PR can serve as a launching point for additional permissions/role based changes for the 2.5 project

* RoleRegistry contract
* call blacklist -> whitelist
* simplified external calling architecture + security
* updated all tests to new system

### CODE REVIEW

File: src/EtherFiNodesManager.sol [hunk 1]
```diff
@@ -68,6 +68,9 @@ contract EtherFiNodesManager is
 
     mapping(address => bool) public eigenLayerOperatingAdmin;
 
+    mapping(address => mapping(bytes4 => bool)) public allowedOperatorCalls;
```
Comment (seongyun-ko) [inline src/EtherFiNodesManager.sol:71]: i believe it was for a separate proposal?

File: src/RoleRegistry.sol [hunk 2]
```diff
@@ -0,0 +1,50 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.24;
+
+import "@openzeppelin-upgradeable/contracts/access/AccessControlUpgradeable.sol";
+import "@openzeppelin-upgradeable/contracts/access/Ownable2StepUpgradeable.sol";
+import "@openzeppelin-upgradeable/contracts/proxy/utils/UUPSUpgradeable.sol";
+
+contract RoleRegistry is AccessControlUpgradeable, UUPSUpgradeable, Ownable2StepUpgradeable {
```
Comment (seongyun-ko) [inline src/RoleRegistry.sol:8]: are you going to add anything more to this contract?

Comment (solipsis) [reply src/RoleRegistry.sol:8]: see updates for blacklist -> whitelist changes. Wanted to ensure it met the requirements of its first real application

File: src/EtherFiNodesManager.sol (hunk 1 above)
Comment (solipsis) [reply src/EtherFiNodesManager.sol]: combined with this update to ensure it actually met the functional requirements of its first intended application

File: src/EtherFiNodesManager.sol [hunk 3]
```diff
@@ -378,45 +449,21 @@ contract EtherFiNodesManager is
...
     // - verifyBalanceUpdates
     // - verifyAndProcessWithdrawals
-    function callEigenPod(uint256[] calldata _validatorIds, bytes[] calldata data) external nonReentrant whenNotPaused onlyEigenLayerOperatingAdmin returns (bytes[] memory returnData) {
-        returnData = new bytes[](_validatorIds.length);
-        for (uint256 i = 0; i < _validatorIds.length; i++) {
-            returnData[i] = IEtherFiNode(etherfiNodeAddress[_validatorIds[i]]).callEigenPod(data[i]);
-        }
-    }
+    function forwardEigenpodCall(uint256[] calldata _validatorIds, bytes[] calldata _data) external nonReentrant whenNotPaused returns (bytes[] memory returnData) {
+        if (!roleRegistry.hasRole(EIGENPOD_CALLER_ROLE, msg.sender)) revert IncorrectRole();
```
Comment (seongyun-ko) [inline src/EtherFiNodesManager.sol:453]: I think we now should check the forward call whitelist on `allowedForwardedEigenpodCalls` within `EtherFiNodesManager` contract?

File: src/EtherFiNode.sol [hunk 4]
```diff
@@ -483,54 +485,55 @@ contract EtherFiNode is IEtherFiNode {
...
-        assembly {
-            selector := mload(add(data, 0x20))
-        }
+    function _verifyEigenPodCall(bytes calldata _data) internal view {
+
+        if (_data.length < 4) revert InvalidForwardedCall();
+        bytes4 selector = bytes4(_data[:4]);
+
+        // can add extra restrictions to specific calls here i.e. checking specific paramaters
+        if (!IEtherFiNodesManager(etherFiNodesManager).allowedForwardedEigenpodCalls(selector)) revert ForwardedCallNotAllowed();
```
Comment (seongyun-ko) [inline src/EtherFiNode.sol:508]: I think we can move all these checks on the forward calls to `EtherFiNodesManager` contract. just need to make suer the forward call are made via `EtherFiNodesManager` contract by putting `onlyEtherFiNodeManagerContract` where needed

File: src/EtherFiNode.sol [hunk 5]
```diff
@@ -483,54 +485,55 @@ contract EtherFiNode is IEtherFiNode {
...
-            selector := mload(add(data, 0x20))
-        }
-        bool allowed = (selector != IDelegationManager.completeQueuedWithdrawal.selector && selector != IDelegationManager.completeQueuedWithdrawals.selector);
-        require (allowed, "NOT_ALLOWED");
+    function _verifyForwardCall(address _to, bytes calldata _data) internal view {
+
+        if (_data.length < 4) revert InvalidForwardedCall();
+        bytes4 selector = bytes4(_data[:4]);
+
+        if (!IEtherFiNodesManager(etherFiNodesManager).allowedForwardedExternalCalls(selector, _to)) revert ForwardedCallNotAllowed();
```
Comment (seongyun-ko) [inline src/EtherFiNode.sol:526]: I think we can move all these checks on the forward calls to `EtherFiNodesManager` contract. just need to make suer the forward call are made via `EtherFiNodesManager` contract by putting `onlyEtherFiNodeManagerContract` where needed

File: src/EtherFiNodesManager.sol [hunk 6]
```diff
@@ -44,13 +46,13 @@ contract EtherFiNodesManager is
     TNFT public tnft;
     BNFT public bnft;
     IAuctionManager public auctionManager;
-    IProtocolRevenueManager public DEPRECATED_protocolRevenueManager;
+    address public DEPRECATED_protocolRevenueManager;
 
     RewardsSplit public stakingRewardsSplit;
-    RewardsSplit public DEPRECATED_protocolRewardsSplit;
+    uint256 public DEPRECATED_protocolRewardsSplit;
```
Comment (seongyun-ko) [inline src/EtherFiNodesManager.sol:52]: nice

**Review by seongyun-ko — CHANGES_REQUESTED** [review]
looks good :)

plz merge the diff here https://github.com/etherfi-protocol/smart-contracts/pull/88

File: src/EtherFiNode.sol (hunk 4 above)
Comment (solipsis) [reply src/EtherFiNode.sol:508]: yep. I can do that. I assumed you wrote it like this originally due to size constraints of the manager contract so I left it. I will move it

Comment (seongyun-ko) [reply src/EtherFiNode.sol:508]: that is very true actually. how should we refactor EFNM to reduce its size :(

Comment (solipsis) [reply src/EtherFiNode.sol:508]: my simplification work this week should in theory delete a lot of code out of EFNM. So I will go ahead and make the change in this PR and work to cut things out of EFNM in my next PR

**Comment by solipsis** [comment]
> looks good :)
> 
> plz merge the diff here #88

done

**Pull request merged**